pkg/controller/master/node/promote_controller.go

```diff
@@ -0,0 +1,511 @@
+	if _, err := h.createPromoteJob(node); err != nil {
+		return nil, err
+	}
+
+	return startedNode, nil
+}
+
+func (h *PromoteHandler) logPromoteEvent(node *corev1.Node, status string) {
+	preStatus := node.Annotations[HarvesterPromoteStatusAnnotationKey]
+	eventType := corev1.EventTypeNormal
+	switch status {
+	case PromoteStatusUnknown, PromoteStatusFailed:
+		eventType = corev1.EventTypeWarning
+	}
+	nodeReference := &corev1.ObjectReference{
+		Name: node.Name,
+		UID:  types.UID(node.Name),
+		Kind: "Node",
+	}
+	h.recorder.Event(nodeReference, eventType,
+		fmt.Sprintf("NodePromote%s", strings.Title(status)),
+		fmt.Sprintf("Node %s promote status change: %s => %s", node.Name, preStatus, status))
+}
+
+// setPromoteStart set node unschedulable and set promote status running.
+func (h *PromoteHandler) setPromoteStart(node *corev1.Node) (*corev1.Node, error) {
+	if node.Annotations[HarvesterPromoteStatusAnnotationKey] == PromoteStatusRunning {
+		return node, nil
+	}
+	h.logPromoteEvent(node, PromoteStatusRunning)
+	toUpdate := node.DeepCopy()
+	toUpdate.Labels[KubeNodeSVCEnableLBLabelKey] = "true"
+	toUpdate.Annotations[HarvesterPromoteStatusAnnotationKey] = PromoteStatusRunning
+	toUpdate.Spec.Unschedulable = true
+	updatedNode, err := h.nodes.Update(toUpdate)
+	return updatedNode, checkError(fmt.Sprintf("setPromoteStart on node %s", node.Name), err)
+}
+
+// setPromoteResult set node schedulable and update promote status if the promote is successful
+func (h *PromoteHandler) setPromoteResult(job *batchv1.Job, node *corev1.Node, status string) (*batchv1.Job, error) {
+	if node.Annotations[HarvesterPromoteStatusAnnotationKey] == status {
+		return job, nil
+	}
+	h.logPromoteEvent(node, status)
+	toUpdate := node.DeepCopy()
+	toUpdate.Annotations[HarvesterPromoteStatusAnnotationKey] = status
+	if status == PromoteStatusComplete {
+		toUpdate.Spec.Unschedulable = false
+	}
+	_, err := h.nodes.Update(toUpdate)
+	return job, checkError(fmt.Sprintf("setPromoteResult to %s on node %s", status, node.Name), err)
+}
+
+type SortByAgeNodeList []*corev1.Node
+
+func (s SortByAgeNodeList) Len() int      { return len(s) }
+func (s SortByAgeNodeList) Swap(i, j int) { s[i], s[j] = s[j], s[i] }
+func (s SortByAgeNodeList) Less(i, j int) bool {
+	// reverse order
+	return !s[i].CreationTimestamp.Before(&s[j].CreationTimestamp)
+}
+
+// selectPromoteNodeList return a list contain nodes need to be promoted
+// If the cluster doesn't need to be promoted, return nil
+func (h *PromoteHandler) selectPromoteNodeList(nodeList []*corev1.Node) ([]*corev1.Node, error) {
+	masterNumber := 0
+	ignoreNodeNumber := 0
+	canBePromoteNodeList := make([]*corev1.Node, 0, len(nodeList))
+	for _, node := range nodeList {
+		if h.shouldBeMaster(node) {
+			masterNumber++
+		} else {
+			if h.isHarvesterNode(node) {
+				if h.isHealthyNode(node) {
+					canBePromoteNodeList = append(canBePromoteNodeList, node)
+				}
+			} else {
+				ignoreNodeNumber++
+			}
+		}
```

Comment:
Looks like nested conditions can be avoided
e.g.
```
if shoudBeMaster{
}else if isHarvesterNode() && isHealthyNode(){
else{
}
```

Reply:
if node is harvester agent node but unhealthy. should not `ignoreNodeNumber++`

Reply:
Should we ignore all nodes that cannot be promoted?
Consider the following case:
1 master 1 healthy 1 unhealthy
masterNumber=1, ignoreNodeNumber=0
specMasterNumber=len(nodeList) - ignoreNodeNumber = 3

But given that there are no enough healthy nodes, do we want to do promote in this case?

Reply:
In this case, the apiserver will be unavailable, three master node is more stable.